contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,470 @@
+
+        // transfer liquidity tokens to erc20 contract
+        UniswapV4ERC20 erc20 = UniswapV4ERC20(poolToERC20[key.toId()]);
+
+        erc20.transferFrom(msg.sender, address(0), liquidity);
+
+        modifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: -int256(liquidity)
+            })
+        );
+
+        // here, all of the necessary liquidity should have been removed, this portion is just to update fees and feeGrowth
+        HookPosition storage position = poolToHookPosition[key.toId()];
+
+        uint128 positionLiquidity = position.liquidity;
+        require(positionLiquidity >= liquidity);
+
+        Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        position.tokensOwed0 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside0LastX128 - position.feeGrowthInside0LastX128,
+                positionLiquidity,
+                FixedPoint128.Q128
+            )
+        );
+        position.tokensOwed1 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside1LastX128 - position.feeGrowthInside1LastX128,
+                positionLiquidity,
+                FixedPoint128.Q128
+            )
+        );
+
+        position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        // subtraction is safe because we checked positionLiquidity is gte liquidity
+        position.liquidity = uint128(positionLiquidity - liquidity);
+    }
+
+    // deploy ERC-20 contract
+    function beforeInitialize(address, IPoolManager.PoolKey calldata key, uint160) external override returns (bytes4) {
+        require(key.tickSpacing == 60, "Tick spacing must be default");
+
+        // deploy erc20 contract
+
+        bytes memory bytecode = type(UniswapV4ERC20).creationCode;
+        bytes32 salt = keccak256(abi.encodePacked(key.toId()));
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        }
+
+        poolToERC20[key.toId()] = poolToken;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function _rebalance(IPoolManager.PoolKey calldata key) internal {
+        HookPosition storage position = poolToHookPosition[key.toId()];
+
+        BalanceDelta balanceDelta = hookModifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: -int256(int128(position.liquidity))
+            })
+        );
+
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
+
+        uint128 liquidity = LiquidityAmounts.getLiquidityForAmounts(
```

Comment:
i dont think you need to do this. liquidity for a full-range position should be min(y / sqrtPrice, x * sqrtPrice). you also can never be out of range in this position, so you dont need to pass sqrtPrice ratios. should save you some expensive calls.

this is because p_b and p_a are functionally infinity and 0, so the liquidity formulas resolve to the one above!

Reply:
ok hmm I did this because the min tick and max tick in this hook are actually a little smaller than the actually set min tick and max tick borders for all v4 pools, since we're using a tick spacing of 60. so I thought it might not be completely exact to the v2 formula